django/db/backends/postgresql/operations.py

```diff
@@ -368,3 +382,33 @@ def on_conflict_suffix_sql(self, fields, on_conflict, update_fields, unique_fiel
             update_fields,
             unique_fields,
         )
+
+
+_positional_param_re = _lazy_re_compile(r"%%|%s")
+
+
+def compose(query, params, context):
```

Comment:
I believe that all this can be done in psycopg 3.1 by using [client-side cursors](https://www.psycopg.org/psycopg3/docs/advanced/cursors.html#client-side-binding-cursors). They were added in psycopg after this functions and its use were written.

...

I have reviewed the use of this function. Probably this is the right abstraction, but it can be probably implemented using `ClientCursor(context).mogrify(query, params)`. `context` would be better called (and typed as) `connection`.

Reply:
Ah yes, I read about that and I am already thinking of using client side cursors completely so we can get on psycopg3 and then fix the remaining server side issues later on.

Reply:
I replaced compose with mogrify in af21e57f19d36ce75a007551bdf49456713e32c3 -- I'd appreciate if you looked over it in case I missed something. I'll mark is as resolved in the meantime, feel free to unresolve if an issue persists.